src/geo/transform.js

```diff
@@ -1172,29 +1193,46 @@ class Transform {
         minmax.min *= elevation.exaggeration();
         minmax.max *= elevation.exaggeration();
-        const top = this.horizonLineFromTop();
-        return [
-            new Point(0, top),
-            new Point(this.width, top),
-            new Point(this.width, this.height),
-            new Point(0, this.height)
-        ].reduce((acc, p) => {
-            return acc
-                .extend(this.coordinateLocation(this.rayIntersectionCoordinate(this.pointRayIntersection(p, minmax.min))))
-                .extend(this.coordinateLocation(this.rayIntersectionCoordinate(this.pointRayIntersection(p, minmax.max))));
-        }, new LngLatBounds());
+        // const top = this.horizonLineFromTop();
+
+        const topLeft = new Point(this._edgeInsets.left, this._edgeInsets.top);
+        const topRight = new Point(this.width - this._edgeInsets.right, this._edgeInsets.top);
+        const bottomRight = new Point(this.width - this._edgeInsets.right, this.height - this._edgeInsets.bottom);
+        const bottomLeft = new Point(this._edgeInsets.left, this.height - this._edgeInsets.bottom);
+
+        // Consider far points at the maximum possible elevation
+        // and near points at the minimum to ensure full coverage.
+        let tl = this.pointCoordinate(topLeft, minmax.min);
+        let tr = this.pointCoordinate(topRight, minmax.min);
+        const br = this.pointCoordinate(bottomRight, minmax.max);
+        const bl = this.pointCoordinate(bottomLeft, minmax.max);
+
+        // Snap points if off the edge of the map.
+        const slope = (p1, p2) => (p2.y - p1.y) / (p2.x - p1.x);
+
+        if (tl.y > 1 && tr.y >= 0) tl = new MercatorCoordinate((1 - bl.y) / slope(bl, tl) + bl.x, 1);
+        else if (tl.y < 0 && tr.y <= 1) tl = new MercatorCoordinate(-bl.y / slope(bl, tl) + bl.x, 0);
+
+        if (tr.y > 1 && tl.y >= 0) tr = new MercatorCoordinate((1 - br.y) / slope(br, tr) + br.x, 1);
+        else if (tr.y < 0 && tl.y <= 1) tr = new MercatorCoordinate(-br.y / slope(br, tr) + br.x, 0);
+
+        return new LngLatBounds()
+            .extend(this.coordinateLocation(tl))
+            .extend(this.coordinateLocation(tr))
+            .extend(this.coordinateLocation(bl))
+            .extend(this.coordinateLocation(br));
+
     }
 
     /**
-     * Returns position of horizon line from the top of the map in pixels. If horizon is not visible, returns 0.
+     * Returns position of horizon line from the top of the map in pixels.
+     * If horizon is not visible, returns 0 by default or a negative value if called with clampToTop = false.
      * @private
      */
     horizonLineFromTop(clampToTop: boolean = true): number {
         // h is height of space above map center to horizon.
         const h = this.height / 2 / Math.tan(this._fov / 2) / Math.tan(Math.max(this._pitch, 0.1)) + this.centerOffset.y;
-        // incorporate 3% of the area above center to account for reduced precision.
-        const horizonEpsilon = 0.03;
```

Comment:
We're moving from 0.03 to 0.1 by using `this._horizonShift`, is there a big visual difference or are we better matching the horizon after this change? It might be nice to have a small debug page to visually test this change, by drawing a line where `horizonLineFromTop` lies in the viewport.

Reply:
This actually doesn't make a visual difference, since the rendering code was already using `_horizonShift` (in `_calcMatrices()`). This change just makes horizon-related functions now return values more consistent with the rendered map. (Functions that call horizonLineFromTop are sampleAverageElevation, coveringTiles, pointCoordinate, pointCoordinate3D, isPointAboveHorizon, isCornerOffEdge( formerly isHorizonVisibleForPoints)).